src/tools/diagnostics.md

```diff
@@ -0,0 +1,377 @@
+---
+title: Diagnostics
+description: Details for diagnostics produced by the Dart analyzer.
+---
+
+This page lists diagnostic messages produced by the Dart analyzer,
+with details about what those messages mean
+and how you can fix your code.
+For more information about the analyzer, see
+[Customizing static analysis](/guides/language/analysis-options).
+
+
+## Glossary
+
+This page uses the following terms.
+
+### Element
+
+[PENDING: definition]
+
+### Potentially non-nullable
+
+A type is _potentially non-nullable_ if it's either explicitly non-nullable or
+if it's a type parameter.
+The latter case is included because
+the actual runtime type might be non-nullable.
+
+
+## List of diagnostics
+
+The analyzer produces the following diagnostics when [PENDING...].
```

Comment:
The best I've come up with so far is

> The analyzer produces the following diagnostics when the code being analyzed
> does not conform to the language specification or contains other errors that
> will cause it to work in unexpected ways.

but it might be better to just write

> The analyzer produces the following diagnostics.

and leave it at that.

Reply:
I like the explanation in the first one. I'd tweak it a bit (of course, I'm me :), so I'd say something more like this:

> The analyzer produces the following diagnostics for code that
> doesn't conform to the language specification or that
> contains other errors that might cause it to work in unexpected ways.

Is "errors" correct? Also, we don't know what they expect, so "might" seemed warranted. I was tempted to say "...errors that might cause it to misbehave." But that seemed a little too anthropomorphic.

Reply:
> Is "errors" correct?

In the "Diagnostic Design Guidelines" I chose "problem", but I'm not sure what will read best to our users. The word "issue" is another possibility. Any synonym is probably fine.

> ... and we avoid "will" ...

Is it OK in contexts where it's accurate, such as when we're describing the runtime semantics of the language ("doing X will cause Y to happen")? I'll try to curb my use of it elsewhere.

Reply:
How about if we avoid the question altogether by saying

> The analyzer produces the following diagnostics for code that doesn't conform to the language specification or that might work in unexpected ways.

Reply:
> > ... and we avoid "will" ...
> 
> Is it OK in contexts where it's accurate, such as when we're describing the runtime semantics of the language ("doing X will cause Y to happen")? I'll try to curb my use of it elsewhere.

In that case, I'd still use "causes" instead of "will cause".